modules/renter/contractor/uptime.go

```diff
@@ -0,0 +1,76 @@
+package contractor
+
+import (
+	"time"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/modules"
+)
+
+// uptimeMinScans is the minimum number of scans required to judge whether a
+// host is offline or not.
+const uptimeMinScans = 3
+
+// uptimeWindow specifies the duration in which host uptime is checked.
+var uptimeWindow = func() time.Duration {
+	switch build.Release {
+	case "dev":
+		return 30 * time.Minute
+	case "standard":
+		return 7 * 24 * time.Hour // 1 week
+	case "testing":
+		return 15 * time.Second
+	}
+	panic("undefined uptimeWindow")
+}()
+
+// isOffline decides whether a host should be considered offline, based on its
+// scan metrics.
+func isOffline(host modules.HostDBEntry) bool {
+	// consider a host offline if:
+	// 1) The host has been scanned at least three times, and
+	// 2) The three most recent scans have all failed, and
+	// 3) The time between the most recent scan and the last successful scan
+	//    (or first scan) is at least uptimeWindow
+	numScans := len(host.ScanHistory)
+	if numScans < uptimeMinScans {
+		// not enough data to make a fair judgment
+		return false
+	}
+	// NOTE: ScanHistory is ordered from oldest-newest
```

Comment:
"I thought of another potential issue: when determining the scan window, we call windowEnd.Sub(windowStart)"

This is exactly the type of thing that makes me really paranoid about that sort of assumption. It's possible that there are also other potential problems, or that there will be other potential problems in the future after we make additional modifications. It opens up a vector to cause inconsistent state, and that inconsistent state could potentially lead to other inconsistent state, which at some point could lead to a more serious bug.

I agree that we can't protect against users intentionally doing bad things to their program, but we should at least make it so that the state will not become inconsistent - instead it should crash or return an error (or be designed so that it can't become inconsistent).